### ISSUE TITLE
Refactor cmake

### ISSUE THREAD

**Issue by pcchen** [Cytnx-dev/Cytnx#473]
I create this issue to document information needed to refactor cmake related material.

Minimum requirements to build cytnx. Here is the incomplete list.

For C++
* boost
* blas/lapack. Effectively, we only support MKL and OpenBLAS. Maybe we only officially support these two.

For python wrapper
* python
* pybind11

For GPU support
* CUDA
* cuTENSOR
* cuQuantum

For OpenMP

**Comment by pcchen** [comment]
CMP0167 https://cmake.org/cmake/help/latest/policy/CMP0167.html

The [FindBoost](https://cmake.org/cmake/help/latest/module/FindBoost.html#module:FindBoost) module is removed in cmake 3.30. 

find_package(Boost CONFIG) finds the upstream package directly, without the find module.

**Comment by pcchen** [comment]
CMP0146 https://cmake.org/cmake/help/latest/policy/CMP0146.html

The [FindCUDA](https://cmake.org/cmake/help/latest/module/FindCUDA.html#module:FindCUDA) module is removed in cmake 3.27.

Examples
* https://cliutils.gitlab.io/modern-cmake/chapters/packages/CUDA.html

**Comment by pcchen** [comment]
OpenMP examples
* https://cliutils.gitlab.io/modern-cmake/chapters/packages/OpenMP.html

**Comment by pcchen** [comment]
CUDA Quantum and CMake[¶](https://nvidia.github.io/cuda-quantum/0.6.0/using/advanced/cmake_app.html#cuda-quantum-and-cmake)

**Comment by pcchen** [comment]
OpenBLAS

If libopenblas is installed in /opt/OpenBLAS/

On Linux, after setting
```
export C_INCLUDE_PATH=/opt/OpenBLAS/include
export CPLUS_INCLUDE_PATH=/opt/OpenBLAS/include
export LD_LIBRARY_PATH=/opt/OpenBLAS/lib
```
cmake can find OpenBLAS and one can compile cytnx.

On MacOS (Apple silicon), still cmake cannot find OpenBLAS.

On Linux and MacOS, one can use `CMAKE_PREFIX_PATH`.
```
CMAKE_PREFIX_PATH=/opt/OpenBLAS cmake ..
```
then cmake can find OpenBLAS. But it will fail to compile because it cannot find `lapacke.h`.
One still have to set the include_path as above. Then it will compile OK.

==> Next thing is to try to use submodules to handle OpenBLAS.

**Comment by IvanaGyro** [comment]
## Status consolidation (checked against current `master`, `a969a66c`)

I agree that #473 should no longer drive a new monolithic “refactor CMake” PR: its requirements have been decomposed into focused issues and PRs.  However, it would be inaccurate to call every item complete.  The all-platform Pixi workspace is still an open PR, and a few CMake discovery/export tasks remain open.

### CMake foundation and policy work

- **Presets / modern package configuration:** PR #579 replaced `Install.sh` with `CMakePresets.json`; PR #809 modernized `CytnxConfig.cmake` and re-finds public imported dependencies; PR #1110 added MSVC CMake portability; and PR #1149 added the usable Windows CUDA presets.
- **CMP0146 / FindCUDA:** resolved in the actual configuration path: current master uses `enable_language(CUDA)` plus `find_package(CUDAToolkit)`, not `FindCUDA`.  #949 remains only to clean up stale legacy variables in the configure-time status output; PR #950 is the broader open CUDA/CMake hardening umbrella.
- **CMP0167 / Boost:** not fully resolved.  Installed-package discovery is config-mode after PR #809, and PR #1123 fixed the CUDA-wheel `CMAKE_PREFIX_PATH` handling plus its Boost pin, but the non-Windows root build still uses bare `find_package(Boost REQUIRED)`.  That direct follow-up is #1122.

### Original dependency checklist

- **Boost:** PR #1057 moved the Linux/macOS wheel-native dependency source to conda-forge; PR #1123 fixes the CUDA wheel’s CMake prefix/Boost compatibility.  The remaining CMake-policy work is #1122; reproducible contributor provisioning is in PR #1150.
- **OpenBLAS / MKL / LAPACK / LAPACKE:** PR #605 established the MKL LP64 direction; PR #800 adopted Morse’s maintained `FindLAPACKE`; PR #1057 provisions conda-forge OpenBLAS/LAPACKE for wheels.  #516 remains the broader LAPACKE discussion.  Relocatable downstream usage requirements are tracked by #1120 and its active implementation PR #1157.
- **Python / pybind11:** PR #755 established the pybind11 3.x direction and PR #1110 uses config-mode Python/pybind11 discovery on MSVC.  #638 remains the test/documentation record; PR #1150 will pin/provision the development-side Python and pybind11 toolchain.
- **CUDA / cuTENSOR / cuQuantum:** PR #875 made CUDA configuration GPU-less-host friendly and raised the CMake floor; PR #993 made the custom finders honest/version-aware; PR #1131 fixed their per-CUDA-major library-layout search; PR #1023 established the CUDA PyPI wheel/runtime path; and PR #1149 provides Windows CUDA presets.  The remaining direct work is spread across #462 and #580 (policy/options), #945/#946/#948/#949/#962 (requirements, finder/rpath/diagnostics), #1129 and #1136 (root-variable and `CMAKE_PREFIX_PATH` behavior), #950 (partly split umbrella), and #1157 (exported SDK usage requirements).  Do not count PR #1130 as a lasting documentation completion: it was reverted by PR #1134.
- **OpenMP:** PR #581 fixed propagation for the static HPTT dependency, and PR #1057 deliberately uses an OpenMP OpenBLAS build to avoid competing thread pools.  #583 remains the broader CI-configuration coverage tracker.

### Pixi / reproducible development environments

- #454 is closed with the decision to use managed packages for external build dependencies rather than make Boost, pybind11, or OpenBLAS external git submodules.
- #1114 is the Windows build/release umbrella.  The closed Windows-only predecessor PR #1117 was rebased into the current all-platform Pixi PR #1150.
- #1140 defines the Linux/macOS extension of that model: one locked, five-platform workspace covering Boost, OpenBLAS/MKL + LAPACKE, Python/pybind11, CUDA/cuTENSOR, contributor tools, and existing CMake presets.
- **PR #1150 is the active implementation** of that proposal.  It is not merged yet, so `pixi.toml`/`pixi.lock` are not currently on master.
- #1158 tracks the required CI smoke coverage for the manifest, lockfile, and Pixi tasks after the workspace lands.

In short: #473 remains a useful historical index, while the dependency-provisioning work now belongs to #1140/PR #1150/#1158 and the remaining CMake correctness work belongs to #1122, #1120/PR #1157, and the CUDA finder/export follow-ups above.  There is no need for a new umbrella implementation under #473, but it should not be closed as fully completed until those active items are resolved or explicitly declined.
